lms/djangoapps/discussion/views.py

```diff
@@ -200,20 +203,66 @@ def inline_discussion(request, course_key, discussion_id):
     except ValueError:
         return HttpResponseServerError("Invalid group_id")
 
-    with function_trace("get_metadata_for_threads"):
-        annotated_content_info = utils.get_metadata_for_threads(course_key, threads, request.user, user_info)
+    # TODO: move all of this into utils.get_metadata_for_threads
+    flattened_threads = []
+    def __flatten(thread):
+        """Turn our tree of threads into a list"""
+        flattened_threads.append(thread)
+        children = thread.get('children', []).append(
+            thread.get('endorsed_responses', [])
+        ).append(
+            thread.get('non_endorsed_responses', [])
+        )
+        for thread in children:
+            __flatten(thread)
+    for thread in threads:
+        __flatten(thread)
+
+    flat_uinfo = {
+        'up': set(user_info['upvoted_ids']),
+        'down': set(user_info['downvoted_ids']),
+        'sub': set(user_info['subscribed_thread_ids']),
+    }
+    user_perms = all_permissions_for_user_in_course(request.user, course_key)
+    course_discussion_settings = get_course_discussion_settings(course_key)
+    course_is_divided = course_discussion_settings.division_scheme is CourseDiscussionSettings.NONE
+    content_usernames = {thread['username'] for thread in threads}
+    content_user_group_ids = {
+        name: get_group_id_for_user(
+            get_user_by_username_or_email(name),
+            course_discussion_settings
+        ) for name in content_usernames
+    }
+    user_group_id = get_group_id_for_user(request.user, course_discussion_settings)
+    annotated_content_info = {
+        thread['id']: {
+            'voted': 'up' if thread['id'] in flat_uinfo['up'] else 'down' if thread['id'] in flat_uinfo['down'] else '',
+            'subscribed': thread['id'] in flat_uinfo['sub'],
+            'ability': get_ability2(
+                thread,
+                request.user,
+                user_perms,
+                course_is_divided,
+                user_group_id,
+                content_user_group_ids[thread['username']]
+            ),
+        } for thread in flattened_threads
+    }
+    # ennd TODO
```

Comment:
I don't think the "flattening" gains us much here; the improvements to group id lookup are solid though